Add plugin: Publish to DEV

stroiman opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/stroiman/obsidian-dev-publish

## Release Checklist
- [x] I have tested the plugin on
  - [ ]  Windows
  - [x]  macOS
  - [ ]  Linux
  - [ ]  Android _(if applicable)_
  - [x]  iOS _(if applicable)_
- [x] My GitHub release contains all required files
  - [x] `main.js`
  - [x] `manifest.json`
  - [ ] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

stroiman commented:
> Thank you for your submission, an automated scan of your plugin code's revealed the following issues:
> 
> ### Optional
> [[1](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/src/medium-gateway.ts#L91)]:Casting to `any` should be avoided as much as possible.
> 
> Do NOT open a new PR for re-validation.

I absolutely agree :)

The cast is part of some validation code, that actually verifies that a type received from an external source conforms to a specification. Is is only _inside_ the implementation of the validation function, and the returned value IS strongly typed, a type that is generated by the actual validation specification.

Normally, I would use a library like io-ts or similar for this type of work, but to keep the plugin code small, I implemented the required behaviour inside the plugin instead, also as I only need a fraction of the functionality provided by such libraries.

joethei commented:
["description": "Write notes in Obsidian, publish and update them as articles on DEV (https://dev.to)",](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/manifest.json#L6)
Don't include "Obsidian" here.

[name: "Publish current note to dev.to",](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/main.ts#L34)
This might be a bit long, in the command pallete it will show as `Publish to DEV: Publish current note to dev.to`, the duplication is not necessary.

[console.log("Stuff published to dev.to");](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/main.ts#L46), [(err) => console.error("Error publishing to dev.to"),](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/main.ts#L48)
This should be shown in a Notice, and not hidden in the dev console.

[// If the plugin hooks up any global DOM events (on parts of the app that doesn't belong to this plugin)](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/main.ts#L59-L68)
Please remove the sample code

[async getFrontMatter(file: TFile) {](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/src/publisher.ts#L36-L40)
To read frontmatter use `this.app.metadataCache.getFileCache(file)?.frontmatter`
`processFrontmatter` should only be used if you are also writing to frontmatter within that.

[async getFrontMatter(file: TFile) {](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/src/publisher.ts#L36-L40)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

stroiman commented:
Thanks for the helpful feedback, e.g. mentioning what to use instead of what I had used.

> ["description": "Write notes in Obsidian, publish and update them as articles on DEV (https://dev.to)",](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/manifest.json#L6) Don't include "Obsidian" here.

Changed to "Publish and update notes as articles on DEV (https://dev.to)"

> [name: "Publish current note to dev.to",](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/main.ts#L34) This might be a bit long, in the command pallete it will show as `Publish to DEV: Publish current note to dev.to`, the duplication is not necessary.

Changed the command name to "Create/update article", so the entire item is shown in the palette as `Publish to DEV: Create/update article`.

I changed "Publish" to "Create/update", as articles are in the current version not "published", but draft, so the use of "publish" was a bit misleading. Also want to emphasize the "update" part, that changes doesn't become new articles.

> [console.log("Stuff published to dev.to");](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/main.ts#L46), [(err) => console.error("Error publishing to dev.to"),](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/main.ts#L48) This should be shown in a Notice, and not hidden in the dev console.

Done. Unsure what is common practice in regards to automatically dismissing the notice. But as I assume that this is a message that the user should not miss, I set it to not automatically dismissing.

> [// If the plugin hooks up any global DOM events (on parts of the app that doesn't belong to this plugin)](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/main.ts#L59-L68) Please remove the sample code

Done (thanks - embarrassing I missed that, though I had deleted all unused sample code).

> [async getFrontMatter(file: TFile) {](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/src/publisher.ts#L36-L40) To read frontmatter use `this.app.metadataCache.getFileCache(file)?.frontmatter` `processFrontmatter` should only be used if you are also writing to frontmatter within that.

Done, thanks for pointing in the right direction.

I was a little unsure how "reliable" this is. The name "cache" seems to indicate that it _may_ be stale, and the fact that it's not async, also indicates that it's not refreshed on read.

> [async getFrontMatter(file: TFile) {](https://github.com/stroiman/obsidian-dev-publish/blob/8e87b333db8b6ba128043fdd4c84f0ad62f41fed/src/publisher.ts#L36-L40) [Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

I'm unsure what this refers to, the link was the link for the previous issue.
The "only" title case in the referenced file, publisher.ts, is the missing heading to use when not found in the note - but that is the title given to the DEV article, where the predominant style is title case (and at any rate, the author would want to change it anyway).

But in main.ts, I had
- "API KEY" - changed to "API key"
- "SECURITY WARNING!", Changed to `sanitizeHTMLToDom("<b>Security ...!</b><br />...")` to bring emphasis to the warning, rather than using upper case. Is there a particular style that you could apply in such a case?

It took a few iterations, but changes are in the 0.8.0
